components/feature/push/src/main/java/mozilla/components/feature/push/AutoPushFeature.kt

```diff
@@ -193,72 +179,30 @@ class AutoPushFeature(
+     * Subscribes and notifies observers about the subscription information if available.
      *
-     * @param type the push message type that you want to be registered.
-     * @param observer the observer that will be notified.
-     * @param owner the lifecycle owner for the observer. Defaults to [ProcessLifecycleOwner].
-     * @param autoPause whether to stop notifying the observer during onPause lifecycle events.
-     * Defaults to false so that messages are always delivered to observers.
+     * @param scope The subscription identifier which usually represents the website's URI.
      */
-    fun registerForPushMessages(
-        type: PushType,
-        observer: Bus.Observer<PushType, String>,
-        owner: LifecycleOwner = ProcessLifecycleOwner.get(),
-        autoPause: Boolean = false
-    ) = messageObserverBus.register(type, observer, owner, autoPause)
-
-    /**
-     * Notifies observers about the subscription information for the push type if available.
-     */
-    fun subscribeForType(type: PushType) {
+    fun subscribe(scope: String, appServerKey: String? = null, onSubscribe: ((AutoPushSubscription) -> Unit) = {}) {
         DeliveryManager.with(connection) {
-            scope.launchAndTry {
-                val sub = subscribe(type.toChannelId()).toPushSubscription()
-                subscriptionObservers.notifyObservers { onSubscriptionAvailable(sub) }
+            coroutineScope.launchAndTry {
+                val sub = subscribe(scope, appServerKey)
+                onSubscribe(sub)
             }
         }
     }
 
     /**
-     * Returns subscription information for the push type if available.
-     *
-     * Implementation notes: We need to connect this to the device constellation so that we update our subscriptions
-     * when notified by FxA. See [#3859][0].
+     * Un-subscribes from a valid subscription.
      *
-     * [0]: https://github.com/mozilla-mobile/android-components/issues/3859
-     */
-    fun unsubscribeForType(type: PushType) {
-        DeliveryManager.with(connection) {
-            scope.launchAndTry {
-                unsubscribe(type.toChannelId())
-            }
-        }
-    }
-
-    /**
-     * Returns all subscription for the push type if available.
+     * @param scope The subscription identifier which usually represents the website's URI.
      */
-    fun subscribeAll() {
+    fun unsubscribe(scope: String, onUnsubscribe: ((Boolean) -> Unit) = {}) {
         DeliveryManager.with(connection) {
-            scope.launchAndTry {
-                PushType.values().forEach { type ->
-                    val sub = subscribe(type.toChannelId()).toPushSubscription()
-                    subscriptionObservers.notifyObservers { onSubscriptionAvailable(sub) }
+            coroutineScope.launchAndTry {
+                val result = unsubscribe(scope)
+                if (result) {
+                    onUnsubscribe(result)
```

Comment:
If there's a possibility of failure for this operation, shouldn't you also have a `onError` callback as part of this API then? For consumers, currently it's a little strange - the block they passed in for `onUnsubscribe` may never get called - and there's a good chance this is what will "pump" consumer's logic forward.

Reply:
Oops, you're right, we shouldn't filter that out.

> shouldn't you also have a onError callback as part of this API then?
We shouldn't need to be explicit calling `onError` since `launchAndTry` will invoke that if we receive an error.